posthog/special_migrations/runner.py

```diff
@@ -0,0 +1,190 @@
+
+    if not migration_definition.is_required():
+        complete_migration(migration_instance)
+        return False
+
+    for service_version_requirement in migration_definition.service_version_requirements:
+        in_range, version = service_version_requirement.is_service_in_accepted_version()
+        if not in_range:
+            process_error(
+                migration_instance,
+                f"Service {service_version_requirement.service} is in version {version}. Expected range: {str(service_version_requirement.supported_version)}.",
+            )
+
+    ok, error = run_migration_healthcheck(migration_instance)
+    if not ok:
+        process_error(migration_instance, error)
+        return False
+
+    dependency_ok, dependency_name = is_migration_dependency_fulfilled(migration_instance.name)
+
+    if not dependency_ok:
+        process_error(migration_instance, f"Could not trigger migration because it depends on {dependency_name}")
+        return False
+
+    reset_special_migration(migration_instance)
+
+    return run_special_migration_next_op(migration_name, migration_instance)
+
+
+def run_special_migration_next_op(
+    migration_name: str, migration_instance: Optional[SpecialMigration] = None, run_all=True
+):
+    if not migration_instance:
+        try:
+            migration_instance = SpecialMigration.objects.get(name=migration_name, status=MigrationStatus.Running)
+        except SpecialMigration.DoesNotExist:
+            return False
+    else:
+        migration_instance.refresh_from_db()
+
+    assert migration_instance is not None
+
+    migration_definition = get_special_migration_definition(migration_name)
+    if migration_instance.current_operation_index > len(migration_definition.operations) - 1:
+        complete_migration(migration_instance)
+        return True
+
+    op = migration_definition.operations[migration_instance.current_operation_index]
+
+    error = None
+    current_query_id = str(UUIDT())
+
+    try:
+        execute_op(op.database, op.sql, op.timeout_seconds, current_query_id)
+        update_special_migration(
+            migration_instance=migration_instance,
+            current_query_id=current_query_id,
+            current_operation_index=migration_instance.current_operation_index + 1,
+        )
+
+    except Exception as e:
+        error = str(e)
+        process_error(migration_instance, error)
+
+    if error:
+        return False
+
+    update_migration_progress(migration_instance)
+
+    # recursively run through all operations
+    if run_all:
+        return run_special_migration_next_op(migration_name, migration_instance)
+
+
+def run_migration_healthcheck(migration_instance: SpecialMigration):
+    return get_special_migration_definition(migration_instance.name).healthcheck()
+
+
+def update_migration_progress(migration_instance: SpecialMigration):
+    # we don't want to interrupt a migration if the progress check fails, hence try without handling exceptions
```

Comment:
Could you also make this a python docstring?
https://www.python.org/dev/peps/pep-0257/